### PR TITLE
Movement speed overhaul

### LINKED ISSUE #11940: Phazon floors don't make you go fast
(WEB REPORT BY: sonixapache REMOTE: 172.93.109.202:7777)

> Revision (Should be above if you're viewing this from ingame!)
> edc7031057929aca2beb276a08cb404fa78d3641
> 
> General description of the issue
> Phazon floors are supposed to make you go faster, but they do not.
> 
> What you expected to happen
> go fast while walking over them
> 
> What actually happened
> go same speed as always while walking over them

### LINKED ISSUE #13902: How would a sane movement speed overhaul work?
HOW IT CURRENTLY WORKS
Your "tally" is how long it takes you to move one square (basically). Think of your effective speed as 1 divided by tally. Your clothes (actually only your shoes and exosuit) and your held items can give you slowdown, which increases your tally. Then your tally is also increased by other misc. shit like pain, hypothermia, your race etc etc.

Your base tally is 0, which makes no sense because that would mean your base speed is infinite speed. Your shoes have a slowdown of -1. This was the oldcoders' idea of how to implement lack of shoes making you slower: by making not-lack of shoes make you faster.
Then all the misc. shit like pain gets applied blah blah and your tally gets capped at 1 if it's under 1.
What does this mean?
Assuming you're a healthy, warm human with shoes, your exosuit's slowdown would need to be higher than 2 to have ANY effect on you. And if it had 3 slowdown, it would make you TWICE as slow. Oh and if you took off your shoes, THEN you would be 3 times as slow as normal. Which is a ridiculously steep increase, the obvious solution would be slowdowns implemented such as "0.3" but those do not exist in the current code.

Already you can see the problems with this stupid tally system:
- Because your base tally is lower than the minimum tally cap, you can have a bunch of slowdown on you and it will have no effect whatsoever until it goes OVER the minimum cap (the reason why hardsuits and galoshes don't slow you down anymore!)
 - Because your base tally is lower than the minimum tally cap, a healthy human is already running at the fastest speed possible!
 - Let's say something has a negative slowdown of -0.5, 0.5 tally is twice as fast as 1 tally, but 2.5 tally is barely any faster than 3 tally. Now let's say it had a positive slowdown of 0.5, 1.5 tally is 50% slower than 1 tally, but 3.5 tally is barely any slower than 3 tally. Because your "perceived" or "effective" speed is 1/tally, it doesn't scale well at all.
- Even if the minimum tally cap was lowered, negative slowdowns would still be really really wonky if they're just a hard number, because of the reason above!

Now there's two ways this mess can go about being tweaked.
We can keep the same system, implement noshoe slowdown in a sane way, make your base tally 1, and make the minimum tally cap something close to 0.
Then, we tweak all the current slowdown amounts to make them much smaller, like 0.3. Now instead of making you TWICE AS FUCKING SLOW, your hardsuit will make you 30% slower. Then we can add stuff that makes you faster, something like a slowdown of -0.1.
It would be significantly better than what we have have now, but it's still shit because slowdown amounts need to be really really small to not cause people go to LIGHTING FAST (the closer to 0 you get the exponentially faster you get), but then, because everything stacks additively, that same negative slowdown that makes naked people REALLY FAST can barely even offset a third of the slowdown from your hardsuit.

The second way I can think of is to implement all slowdowns, positive and negative, MULTIPLICATIVELY.
So for example, if you're slow as heck from pain (tally 7 or something), and you take some drug that makes you go 10% faster, your final tally would be 1*7*0.9 = 6.3. Or if you're healthy, the drug makes you go at 1*0.9 = 0.9 tally.
The whole point is that it will always make you precisely 10% faster than before. Now we can have multiple sources of negative slowdown, because even if someone stacks them all together, they stack multiplicatively so that it's not terribly broken.

### LINKED ISSUE #13903: Assuming a sane movement speed system was implemented, how should things affect your speed?
Immediate thoughts:

Should hardsuits slow you down again? (trick question they should)
Should galoshes/clown shoes slow you down as they were meant to?

Hyperzine: Junkie gofast drug, or muscle stimulant?
This one's tricky. Historically, hyperzine has always been used to bring you TO your "top speed", not INCREASE your top speed. Obviously, a higher top speed is a very strong thing because then you can outrun everyone else. Personally, I think hyperzine should continue to be "a thing you take when you have a hardsuit so you can move like you didn't have a hardsuit", and a chemical that increases your top speed should be added separately.

Hyperzine+magboots? I know Nuke opened an issue on this but it got bandwagoned pretty hard, I want to see honest opinions here

Should hardsuits slow you down to HALF YOUR FUCKING SPEED (as they were before) or be something like 30% slowdown?

Currently walking is 7 times slower than running. Should that be upped a little?

### PR DESCRIPTION
**Pull request by 9600bauds**
Overhaul of movement speed to fix most things having a slowdown that didn't actually work, and inability for things to make you go faster than base speed.
Effective changes:
Light hardsuits have 40% slowdown (smallest slowdown the new system can handle by itself, please bear in mind that for YEARS the slowest functioning slowdown we've had was 100% slowdown I.E. makes you twice as fucking slow)
Some hardsuits have 80% slowdown
The only hardsuits that previously had slowdown now have a slowdown of 100%, which is what they previously effectively had
Not wearing shoes is 40% slowdown (as it was meant to be a 100% slowdown before)
Galoshes are a 40% slowdown (as they were meant to be a 100% slowdown before)
Removed the tiny slowdown from clown shoes as it practically never worked
Phazon floors make you go 25%-30% faster (from acting like space tiles and just making you always go at base speed instead)
Run mutation makes you go 20%-30% faster (from acting like hyperzine)
Nuka cola makes you go 20%-30% faster (from acting like hyperzine)
Painkillers ignore pain slowdown
Walking is 400% slowdown (from 700%)

About 90% tested

Fixes #11940
Closes #13902
Closes #13903

:cl:
 * experimental: Overhauled the way movement speed is calculated, because the previous system was broken as fuck and didn't let you move any faster than base speed. This means:
 * bugfix: Fixed almost all worn items not having a functioning slowdown like they were supposed to, including hardsuits, galoshes, not wearing shoes, and maybe more.
 * tweak: Phazon floors, the 'run' genetics mutation, and bomberman skate bonuses were re-implemented in a way that ACTUALLY makes you go faster. Nuka-Cola also makes you go faster, as a way of testing the new system.
 * tweak: Made walking a little bit faster, since it previously made you 7 times slower. All speed statistics are not final and open to feedback.

### CODE REVIEW

**Comment by jknpj** [comment]
oh boy

**Comment by 9600bauds** [comment]
OK THINGS THAT STILL NEEDED DECIDING

1) Hyperzine: Makes you resist pain slowdown, or just slowdown from equipped items?
2) Scooting along the outer hull of the station: Removes all slowdown, or muh immulsioning "yes your bulky suit and broken arm are making you crawl along the solar grilles slower"?
3) Clown shoes, galoshes: slowdown or no slowdown?
4) CE magboots were intended to have a slight slowdown: should they?

**Comment by Intigracy** [comment]
I think

Pain too
Space should be speedy
Clown shoes yes, adv clown shoes shouldn't/could speed you up, and yes in galoshes
No, they're the pinnacle of technology

**Comment by yclatious** [comment]
Keep hyperzine as no slowdown, pain is optional.
No slowdown, you are in space and there is no gravity, nothing is making you move slower.
Clown shoes with slowdown are a deathsentence to the honkies, galoshes are fine with it because no slip.
They are already meant to be the latest in magboot tech, Id say ditch the slowdown.

**Comment by 9600bauds** [comment]
also SHOULD FAT PEOPLE WALK LIKE SKINNY PEOPLE THROUGH THE MAGIC OF HYPERZINE.

**Comment by SysadminJeroen** [comment]
fuck

**Comment by SysadminJeroen** [comment]
@9600bauds YES.

**Comment by SonixApache** [comment]
@9600bauds 

**Hyperzine**: Remove equipment slowdown
**Space movement**: Immulsions uber alles, although being able to speed up in space would be a GODSEND
**Clown shoes, galoshes**: slowdown
**CE magboots**: Go slow, bro.
**Fat hyperzine**: cardiac arrest

**Comment by gbasood** [comment]
@9600bauds yeah

**Comment by 9600bauds** [comment]
Is that like "fuck yes" or

**Comment by Intigracy** [comment]
I guess I'm alone thinking fat fucks should be slow

**Comment by MasterofNothing** [comment]
@9600bauds 

Please don't add a slowdown to clown shoes. I always have way too much fun watching clowns escape sec. It'd make them pretty helpless.

The rest sounds good.

**Comment by gbasood** [comment]
I thought clown shoes already had a slowdown

**Comment by Fruktlimpa** [comment]
Hyperzine: Remove equipment slowdown
Space: No slowdown
Clown shoes, galoshes: No slowdown for clown, slight slowdown for galoshes
CE Magboots: No slowdown, supposed to be top tier tech.

**Comment by Probe1** [comment]
Looks good from a balance perspective to me.

**Comment by AI-Is-Rogue** [comment]
I think there should be a slowdown in space when magboots are on, because you are sticking to the hull as you do to floor tiles when they're on. Also it would make it easier for me to navigate the exterior of the space station and not accidentally jump off into space every few minutes.

**Comment by 9600bauds** [comment]
Here's how it used to work: Through spaghetticode, you had 1 slowdown resist. Thanks to inexplicable ancient code, shoes gave you 1 further slowdown resist (except galoshes and clownshoes which had 0).
So, technically, clown shoes did have a RELATIVE slowdown, but it only ever mattered when wearing one of the superslow white softsuits or when you had a broken leg or something.

**Comment by 9600bauds** [comment]
~ignore these they were outdated~

**Comment by 9600bauds** [comment]
ok time for emojicracy

**Comment by 9600bauds** [comment]
hyperzine removes fatness slowdown :+1:/:-1: this comment

edit: Will change hyperzine not enabling fatties and the wounded in a separate PR with the intention of not bundling too many balance changes in one PR

**Comment by 9600bauds** [comment]
hyperzine removes pain slowdown (painkillers also replace pain slowdown now) :+1:/:-1: this comment

edit: Will change hyperzine not enabling fatties and the wounded in a separate PR with the intention of not bundling too many balance changes in one PR

**Comment by 9600bauds** [comment]
clown shoes have 20% slowdown :+1:/:-1: this comment

edit: removed clown shoe slowdown already

**Comment by 9600bauds** [comment]
moving alongside space such as crawling along grilles from solars removes all slowdowns be it pain encumberment etc :+1:/:-1: this comment

**Comment by 9600bauds** [comment]
:+1:/:-1: this comment just do it

**Comment by Lokiamis** [comment]
Have clown shoes give a boost of speed whenever the clown takes brute to encourage beatings and self-harm

**Comment by kilozombie** [comment]
Coffee and brawndo giving 5% and 10% speed bonuses respectively 👍/👎

**Comment by 9600bauds** [comment]
I'm not actually sure if percentages as low as 20% ACTUALLY work, I'll need to test this some more
I might have to up the base hardsuit slowdown to 40% or something before they actually matter

**Comment by 9600bauds** [comment]
Well, after testing, it turns out the human base speed is so fast the server doesn't tick fast enough to make a 30% slowdown matter.
I mean in hindsight this is probably why it didn't work this way back when there was a lower tickrate

so, time to improvise

File: code/modules/clothing/spacesuits/captain.dm [hunk 1]
```diff
@@ -40,7 +40,7 @@
 	pressure_resistance = 200 * ONE_ATMOSPHERE
 	body_parts_covered = ARMS|LEGS|FULL_TORSO|FEET
 	allowed = list(/obj/item/weapon/tank/emergency_oxygen, /obj/item/device/flashlight,/obj/item/weapon/gun/energy, /obj/item/weapon/gun/projectile, /obj/item/ammo_storage, /obj/item/ammo_casing, /obj/item/weapon/melee/baton,/obj/item/weapon/handcuffs,/obj/item/weapon/tank/emergency_nitrogen)
-	slowdown = 1.5
+	slowdown = HARDSUIT_SLOWDOWN_HIGH
 	armor = list(melee = 65, bullet = 50, laser = 50, energy = 25, bomb = 50, bio = 100, rad = 50)
 	siemens_coefficient = 0.7
 	heat_conductivity = SPACESUIT_HEAT_CONDUCTIVITY
```
Comment (CptWad) [inline code/modules/clothing/spacesuits/captain.dm]: Should probably add values to Centcomm armor here and shit like Thunderdome armor.

File: code/modules/clothing/suits/armor.dm [hunk 2]
```diff
@@ -41,7 +41,7 @@
 	icon_state = "riot"
 	item_state = "swat_suit"
 	body_parts_covered = ARMS|LEGS|FULL_TORSO|FEET|HANDS|IGNORE_INV
-	slowdown = 1
+	slowdown = HARDSUIT_SLOWDOWN_LOW
```
Comment (CptWad) [inline code/modules/clothing/suits/armor.dm]: Should be medium due to how ROBUST it is

File: code/modules/clothing/spacesuits/captain.dm [hunk 3]
```diff
@@ -40,7 +40,7 @@
 	pressure_resistance = 200 * ONE_ATMOSPHERE
 	body_parts_covered = ARMS|LEGS|FULL_TORSO|FEET
 	allowed = list(/obj/item/weapon/tank/emergency_oxygen, /obj/item/device/flashlight,/obj/item/weapon/gun/energy, /obj/item/weapon/gun/projectile, /obj/item/ammo_storage, /obj/item/ammo_casing, /obj/item/weapon/melee/baton,/obj/item/weapon/handcuffs,/obj/item/weapon/tank/emergency_nitrogen)
-	slowdown = 1.5
+	slowdown = HARDSUIT_SLOWDOWN_HIGH
```
Comment (CptWad) [inline code/modules/clothing/spacesuits/captain.dm]: Medium please

**Comment by 9600bauds** [comment]
Testing shows:
Any slowdown smaller than 33% is effectively identical to no slowdown unless combined with another slowdown
Any negative slowdown smaller than -33% is effectively identical to -33% unless combined with another slowdown
This is because the ticklag in the serb is 0.33 and that ain't no coincidence

So no tiny coffee speed boost sadly, serb just can't do it

**Comment by kilozombie** [comment]
Damn.
Thanks for trying man.

File: code/modules/clothing/spacesuits/captain.dm (hunk 3 above)
Comment (9600bauds) [reply code/modules/clothing/spacesuits/captain.dm:43]: Medium is essentially low sadly

**Comment by 9600bauds** [comment]
Will change hyperzine not enabling fatties and the wounded in a separate PR with the intention of not bundling too many balance changes in one PR

**Comment by Kurfursten** [comment]
>light hardsuits slow you 40%, which is the smallest the system can handle
>galoshes have 20% slowdown

Really makes me think

**Comment by 9600bauds** [comment]
Thanks for reminding me to update the OP

**Comment by 9600bauds** [comment]
I mean, thinking about it some more.
If you really really wanted a way to implement 20% slowdown, you could do something like...
0.2 is 60% of 0.33, so, add 1 tick to your movement delay 60% of the time
And that would result in a 20% slowdown on AVERAGE but you would have slightly irregular movement, who knows, who really cares

**Comment by NigglyWiggly** [comment]
>Those commit names




healfuck
minefuck
buildfuck

**Comment by kilozombie** [comment]
whatfuck

**Comment by aceedex** [comment]
I do not like this

**Comment by Probe1** [comment]
Why?

**Comment by Blithering** [comment]
The tiny speed boosts from coffee and brawndo might be worth keeping even if they have no tangible benefit if they still count when calculating speed. You never know when you might need an extra 10% to keep you above a threshold.

**Comment by yackemflam** [comment]
Wait what? Why would not wearing shoes give slowdown?
Wearing shoes should give you a SLIGHT slowdown compared to not wearing shoes.
Go outside. Run barefoot vs run with sneakers on.
Barefoot running is faster but hurts because lol concrete

**Comment by SonixApache** [comment]
I run faster with shoes than without 
your argument is invalid you fucking hobbit

**Comment by Kurfursten** [comment]
>Barefoot running is faster

If this were true, every short-distant sprint sporting event would be performed barefoot

**Comment by yackemflam** [comment]
@Kurfursten 
Because running barefoot is painful at first
Studies have shown again and again that running barefoot is faster
Atheletes just wear shoes since running barefoot on a hard surface isnt good for the legs

**Comment by yackemflam** [comment]
Also to note that sprinters wear extremely light shoes

**Comment by Lokiamis** [comment]
it's true I run barefoot and with my arms flailing behind me like sonic and naruto for max speed

**Comment by Shadowmech88** [comment]
>Sonic is the fastest
>Sonic wears shoes

Checkmate.

**Comment by Zth--** [comment]
What about Flash? He wears galoshes

**Comment by Kurfursten** [comment]
I spent the morning reading scholarly journal articles on barefoot running versus what health scientists refer to as "shod" running. There is no evidence that either is faster, prevailing theory is that barefoot running reduces injury because present-day shoe designs encourage striking the heel on the ground. I should note that scientific interest in barefoot running is only about 4 years old. If it interests anyone, running in a practical environment (somewhere with hazards on the floor, for the purposes of our study, poker chips) caused barefoot runners to be more distracted and more stressed because they needed to carefully avoid stepping on obstacles.

From a game design perspective it doesn't matter what is chosen, though. Present-day shoes were developed before sports medicine existed as a discipline and long before barefoot running studies, meanwhile Space Station 13 is set centuries into the future.

**Comment by Lokiamis** [comment]
hi

**Comment by eplgr** [comment]
merge when

**Comment by Exxion** [comment]
Conflicts
@9600bauds whaen

**Comment by Exxion** [comment]
WHAEN

**Comment by BobdaBiscuit** [comment]
never ever

**Comment by Lokiamis** [comment]
every day until you fix conflicts

**Comment by Intigracy** [comment]
https://github.com/vgstation-coders/vgstation13/pull/14840

**Pull request closed without merging**